hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkTableChecker.java

```diff
@@ -148,32 +148,62 @@ public void setTableDataSet(DataSet tableDataSet) {
-                .filter(new DuplicateIndexFilter());
+    private List<String> checkDuplicateIndexes(PairDataSet table, PairDataSet index) throws StandardException {
+        try {
+            SpliceSpark.pushScope(String.format("Check duplicates in index %s.%s", schemaName, indexName));
+            JavaPairRDD duplicateIndexRdd = ((SparkPairDataSet) index).rdd
+                    .combineByKey(new CreateCombiner(), new MergeValue(), new MergeCombiners())
+                    .filter(new DuplicateIndexFilter());
 
-        long count = 0;
-        JavaFutureAction<List> futureAction = duplicateIndexRdd.takeAsync(maxCheckTableError);
-        List<Tuple2<String, List<byte[]>>> result = futureAction.get();
-        for(Tuple2<String, List<byte[]>> tuple : result) {
-            List<byte[]> keys = tuple._2;
-            for (byte[] key : keys) {
-                indexKeyDecoder.set(key, 0, key.length);
-                indexKeyDecoder.decode(indexKeyTemplate);
-                messages.add(indexKeyTemplate.getClone().toString() + "=>" + tuple._1);
-                count++;
+            JavaPairRDD joinedRdd = duplicateIndexRdd
+                    .join(((SparkPairDataSet) table).rdd);
+
+            JavaRDD duplicateIndex = joinedRdd
+                    .mapPartitions(new SparkFlatMapFunction<>(new DeleteDuplicateIndexFunction<>(conglomerate, txn, tentativeIndex, baseColumnMap, fix)));
+
+            Iterator it = duplicateIndex.toLocalIterator();
+
+            if (fix) {
+                return fixDuplicateIndexes(it);
+            } else {
+                return reportDuplicateIndexes(it);
             }
+        }catch (Exception e) {
+            throw StandardException.plainWrapException(e);
+        }
+        finally {
+            SpliceSpark.popScope();
         }
-        if (count >= maxCheckTableError) {
-            count = duplicateIndexRdd.count();
-            messages.add("...");
+    }
+
+
+    private List<String> fixDuplicateIndexes(Iterator it) {
+        List<String> messages = Lists.newLinkedList();
+
+        long count = 0;
+        while (it.hasNext()) {
+            Tuple2<String, Tuple2<byte[], ExecRow>> t = (Tuple2<String, Tuple2<byte[], ExecRow>>)it.next();
+            messages.add(t._2._2 + "=>" + t._1);
+            count++;
         }
+        messages.add(0,String.format("Deleted the following %d duplicate indexes", count));
+        return messages;
+    }
+
+    private List<String> reportDuplicateIndexes(Iterator it) throws StandardException, InterruptedException, ExecutionException {
+        List<String> messages = Lists.newLinkedList();
 
+        long count = 0;
+        while (it.hasNext()) {
+            Tuple2<String, Tuple2<byte[], ExecRow>> t = (Tuple2<String, Tuple2<byte[], ExecRow>>)it.next();
+            messages.add(t._2._2 + "=>" + t._1);
+            if (count >= maxCheckTableError) {
+                messages.add("...");
+                break;
```

Comment:
This was intentional to avoid counting exact number. However, my second thought is that it is rarely a very expensive operations, so I will revert back